### ISSUE TITLE
Special preprocessing for some annotators

### ISSUE THREAD

**Issue by hiroshinoji** [mynlp/jigg#28]
It is known that in JUMAN (as well as Mecab?) hankaku numbers are not recommended. I'm not sure, but one option may be to preprocess the text appropriately in those annotators.

Another related issue I found is that Mecab ignores half spaces in the sentence, while JUMAN does not. Below is the examples:

``` sh
$ echo あ い | juman
あ あ あ 感動詞 12 * 0 * 0 * 0 "代表表記:あ/あ"
  \  \  特殊 1 空白 6 * 0 * 0 NIL
い い いい 形容詞 3 * 0 イ形容詞イ段 19 文語基本形 18 "代表表記:良い/よい 反義:形容詞:悪い/わるい"
EOS

$ echo あ い | mecab
あ フィラー,*,*,*,*,*,あ,ア,ア
い 名詞,一般,*,*,*,*,い,イ,イ
EOS
```

In the current implementation, JumanAnnotator cannot handle an input with half space correctly, since the analyzed token line for a half space (' ') is different from other tokens. One way to solve this is to add preprocessing to remove half spaces from the input in JumanAnnotator. Another way is to preserve the original JUMAN output, and modify the implementation to correctly parse the line for a half space token.

**Comment by hiroshinoji** [comment]
Current behavior of Juman is apparently a bug, so fixed: 5ee95310f76910c37c392a2d0a9777be46a2f2e1

Test: https://github.com/mynlp/jigg/commit/5ee95310f76910c37c392a2d0a9777be46a2f2e1#diff-9bd01309ea508409c8aecb7c36ff95e4R83

**Comment by hiroshinoji** [comment]
This page says KNP recommend half spaces to be replaced with zenkaku spaces.
http://nlp.ist.i.kyoto-u.ac.jp/index.php?KNP%2FFAQ

**Comment by hiroshinoji** [comment]
In fact, KNP cannot process half spaces at all:

``` sh
$ echo a b | juman | knp -tab
;; Invalid input (4 items)<  \  \  特殊 1 空白 6 * 0 * 0 NIL
> !
# S-ID:1 KNP:4.12-CF1.1 DATE:2016/02/17 SCORE:0.00000 ERROR:Cannot make mrph
EOS
```

**Comment by hiroshinoji** [comment]
The above problem for half spaces in KNP is resolved in ebee5efda7de6494d8ee1c5c4a1fda5d02382a24.

Now juman supports `-juman.normalize` option, which is true in default. If this is true, juman analyzes on a normalized text, which transforms all half space characters (except hankaku-kanas) into full space characters. Each `form` or `lemma` in `<token>` given by juman is now based on this normalized text. KNP eats this result, so the problem can be avoided.

Though each `form` may not be different from the original token in the given input, the original form can be obtained using new attributes: `characterOffsetBegin` and `characterOffsetEnd`. Here is an example:

``` scala
val s = (sentences \ "sentence")(0) // sentence node
val t = (s \ "token")(0) // first token
val normalizedForm = t \@ "form"
val originalForm = XMLUtil.text(s) substring (t \@ "characterOffsetBegin", t \@ "characterOffsetEnd")
```

**Comment by hiroshinoji** [comment]
Mecab ignores and does not output space characters. We handle this issue by supporting offset attributes (see #31).
